mkosi/sandbox.py

```diff
@@ -438,6 +443,67 @@ def is_relative_to(one: str, two: str) -> bool:
     return os.path.commonpath((one, two)) == two
 
 
+def pack_file_descriptors() -> int:
+    fds = []
+
+    with os.scandir("/proc/self/fd") as it:
+        for d in it:
+            if not d.is_symlink() and (d.is_file() or d.is_dir()):
+                continue
+
+            try:
+                fd = int(d.name)
+            except ValueError:
+                continue
+
+            if fd < SD_LISTEN_FDS_START:
+                continue
+
+            fds.append(fd)
+
+    # os.scandir() either opens a file descriptor to the given path or dups the given file descriptor. Either
+    # way, there will be an extra file descriptor in the fds array that's not valid anymore now, so find out
+    # which one and drop it.
+    fds = [fd for fd in fds if libc.fcntl(fd, F_GETFD, 0) >= 0]
```

Comment:
Since I just looked at the scandir docs, it mentions that since 3.7 file descriptors are supported, so maybe we could open `/proc/self/fd` ourselves and then right away which fd not to add?

Reply:
@behrmann No that doesn't work, because scandir() will dup that file descriptor so we just move the problem